pkg/net/libp2p/libp2p_transport_test.go

```diff
@@ -598,6 +598,51 @@ func TestLibp2p_OneConnectionInProgress(t *testing.T) {
 	m.testNeverMoreThanOneConnectionInProgress(nodeA)
 }
 
+// Test that when a node fails to connect first time it will be trying to connect.
+func TestLibp2p_OpeningConnectionAfterFail(t *testing.T) {
+	logger := logging.ConsoleDebugLogger
+
+	nodeA := types.NodeID("a")
+	nodeB := types.NodeID("b")
+
+	m := newMockLibp2pCommunication(t, DefaultParams(), []types.NodeID{nodeA, nodeB}, logger)
+
+	a := m.transports[nodeA]
+	b := m.transports[nodeB]
+
+	bBadAddr, err := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/127.0.0.1/tcp/100/p2p/%s", b.host.ID()))
+	require.NoError(t, err)
+
+	require.Equal(t, nodeA, a.ownID)
+	require.Equal(t, nodeB, b.ownID)
+
+	initialNodes := m.Membership(nodeA)
+	initialNodes[nodeB] = bBadAddr
+	t.Log(initialNodes)
+
+	t.Log(">>> connecting to a failed node")
+	err = a.Start()
+	require.NoError(t, err)
+	a.Connect(initialNodes)
+	m.testEventuallyNotConnected(nodeA, nodeB)
```

Comment:
Not sure `EventuallyNotConnected` is what we want here. As I understand it, this test trivially succeeds even if the nodes eventually connect, if it happens to poll the state before the connection is established.

Reply:
The test works as follows: node A tries to connect to node B. node A is connecting to the close port of node B. We check that they are not connected (they never were connected). Then A is trying to connect to the open port of node B. And we check that nodes A succeeds.

Reply:
What I wanted to say is the following: Calling `Connect` (that starts connecting in the background) and immediately testing whether the node is *not* connected creates a form of a race. Even if the node actually did connect one millisecond after you call `testEventuallyNotConnected`, the test would still succeed (an probably should fail)

Reply:
I see. To fix that I explicitly specified parameters for `openStream`: no retries, 1 sec connecting timeout, and no delay.
`testNeverConnected` checks connection during 5 sec, `openStream` establishes a connection during 1 sec.

Reply:
I think even then the problem persists. What if the test succeeds even before the first connection attempt? Since connection happens in the background, this is well possible...